test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -710,3 +790,217 @@ func TestBlockingOwnerRefDoesBlock(t *testing.T) {
+	if err != nil {
+		t.Fatalf("failed to create dependent resource %q: %v", dependent.GetName(), err)
+	}
+	t.Logf("created dependent resource %q", dependent.GetName())
+
+	// Delete the owner.
+	foreground := metav1.DeletePropagationForeground
+	err = resourceClient.Delete(owner.GetName(), &metav1.DeleteOptions{PropagationPolicy: &foreground})
+	if err != nil {
+		t.Fatalf("failed to delete owner resource %q: %v", owner.GetName(), err)
+	}
+
+	// Ensure the dependent is deleted.
+	if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+		_, err := resourceClient.Get(dependent.GetName(), metav1.GetOptions{})
+		return errors.IsNotFound(err), nil
+	}); err != nil {
+		t.Logf("owner: %#v", owner)
+		t.Logf("dependent: %#v", dependent)
+		t.Fatalf("failed waiting for dependent resource %q to be deleted", dependent.GetName())
+	}
+
+	// Ensure the owner is deleted.
+	if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+		_, err := resourceClient.Get(owner.GetName(), metav1.GetOptions{})
+		return errors.IsNotFound(err), nil
+	}); err != nil {
+		t.Fatalf("failed waiting for owner resource %q to be deleted", owner.GetName())
+	}
+}
+
+// TestMixedRelationships ensures that owner/dependent relationships work
+// between core and custom resources.
+//
+// TODO: Consider how this could be represented with table-style tests (e.g. a
+// before/after expected object graph given a delete operation targetting a
+// specific node in the before graph with certain delete options).
+func TestMixedRelationships(t *testing.T) {
+	ctx := setup(t, 5)
+	defer ctx.tearDown()
+
+	clientSet, apiExtensionClient, clientPool := ctx.clientSet, ctx.apiExtensionClient, ctx.clientPool
+
+	ns := createNamespaceOrDie("crd-mixed", clientSet, t)
+
+	configMapClient := clientSet.Core().ConfigMaps(ns.Name)
+
+	definition, resourceClient := createRandomCustomResourceDefinition(t, apiExtensionClient, clientPool, ns.Name)
+
+	// Create a custom owner resource.
+	customOwner, err := resourceClient.Create(newCRDInstance(definition, ns.Name, names.SimpleNameGenerator.GenerateName("owner")))
+	if err != nil {
+		t.Fatalf("failed to create owner: %v", err)
+	}
+	t.Logf("created custom owner %q", customOwner.GetName())
+
+	// Create a core dependent resource.
+	coreDependent := newConfigMap(ns.Name, names.SimpleNameGenerator.GenerateName("dependent"))
+	link(t, customOwner, coreDependent)
+	coreDependent, err = configMapClient.Create(coreDependent)
+	if err != nil {
+		t.Fatalf("failed to create dependent: %v", err)
+	}
+	t.Logf("created core dependent %q", coreDependent.GetName())
+
+	// Create a core owner resource.
+	coreOwner, err := configMapClient.Create(newConfigMap(ns.Name, names.SimpleNameGenerator.GenerateName("owner")))
+	if err != nil {
+		t.Fatalf("failed to create owner: %v", err)
+	}
+	t.Logf("created core owner %q: %#v", coreOwner.GetName(), coreOwner)
+	// TODO: why are these empty following Create?
```

Comment:
The codecs reset the TypeMeta. It's like this because the clients used to convert objects between external and internal version, it would be awkward if the TypeMeta.Version were set to "v1" for an object of the internal version. Because we now rarely do conversion, IMO we don't need to reset TypeMeta anymore.